src/twisted/internet/test/test_endpoints.py

```diff
@@ -3259,6 +3259,41 @@ def test_sslNoTrailingNewlinePem(self):
         ctx = server._sslContextFactory.getContext()
         self.assertIsInstance(ctx, ContextType)
 
+    @skipIf(skipSSL, skipSSLReason)
+    def test_sslWithCustomCipher(self):
+        """
+        A cipher list is supported, using the OpenSSL format.
+        The colon (:) from the OpenSSL format is replaced with a comma (,).
+        Limitations to the test -> Not checked the ciphers returned.
+        """
+        reactor = object()
+        server = endpoints.serverFromString(
+            reactor,
+            "ssl:1234:privateKey=%s:"
+            "certKey=%s:cipher=ALL,!ADH,@STRENGTH,+RSA,-DSA,SHA1+DES"
+            % (escapedPEMPathName, escapedPEMPathName),
+        )
+
+        self.assertIsInstance(server, endpoints.SSL4ServerEndpoint)
+        ctx = server._sslContextFactory.getContext()
+        self.assertIsInstance(ctx, ContextType)
+
+    @skipIf(skipSSL, skipSSLReason)
+    def test_sslInvalidCustomCipher(self):
+        """
+        A cipher list is supported, using the OpenSSL format.
+        The colon (:) from the OpenSSL format is replaced with a comma (,).
+        """
+        reactor = object()
+        with self.assertRaises(Exception) as context:
+            endpoints.serverFromString(
+                reactor,
+                "ssl:1234:privateKey=%s:"
+                "certKey=%s:cipher=bla, blah"
+                % (escapedPEMPathName, escapedPEMPathName),
+            )
+            self.assertTrue("Invalid cipher list passed" in str(context.exception))
```

Comment:
This line is never called as the previous lines raises an exception.

This is why is best to avoid using the context manager

```python
exception = self.assertRaises(
    SomeException,
    endpoints.serverFromString,
    reactor=reactor,
    ...
)
```

Reply:
It's best to folow TDD

Write the test first, run the test and see if fail... then fix the code.

If you write the test first and it passes from the first run, it means that the test is not valid.

https://docs.twisted.org/en/stable/development/test-standard.html

Reply:
[#e9308d](https://github.com/twisted/twisted/pull/12137/commits/6e30f9f83d118633e7be1c489d00073549c1b285)